src/phoneField/index.tsx

```diff
@@ -0,0 +1,179 @@
+  value: object
+}
+interface PhoneFieldProps  {
+  readonly name: string
+  readonly id?: string
+  readonly className?: string
+  readonly defaultPrefix?: string
+  readonly prefixLabel?: string
+  readonly title?: string
+  readonly ariaLabelledBy?: string
+  readonly placeholder?: string
+  readonly defaultValue?: string
+  readonly countriesWhiteList?: string[]
+  readonly customCountriesName?: customCountriesName
+  readonly isInline?: boolean
+  readonly onChange?: (obj: PhoneFieldOnChangeParameters) => void
+}
+
+interface PhoneFieldState {
+  countriesData: object[]
+  countriesWhiteList: string[]
+  completePhoneNumber: string
+  // Make the type of [key: string]: string etc... assignable for setState()
+  [key: string]: any
+}
+
+const filterIso2 = (countriesList: string[]) => countryTelData.allCountries.filter(
+  country => countriesList.includes(country.iso2)
+)
+
+const findIso2 = (countryDefault: string) => countryTelData.allCountries.find(
+  country => country.iso2 === countryDefault
+) || {}
+
+const mapCountriesPhoneData = (
+  countriesData: allCountriesPhoneData[], countriesName: customCountriesName
+  ):mappedCountryPhoneData[] => countriesData.map(countryData => ({
+    value: `+${countryData.dialCode}`,
+    label: `${!isEmpty(countriesName[countryData.iso2]) ?
+      countriesName[countryData.iso2] : countryData.name} +(${countryData.dialCode})`,
+  })
+)
+
+/* Alphabetically sorted */
+const sortCountriesPhoneData = (
+  mappedCountryPhoneData: mappedCountryPhoneData[]
+  ): mappedCountryPhoneData[] => mappedCountryPhoneData.sort(
+    (a, b) => {
+      if(a.label < b.label) return -1
+      if(a.label > b.label) return 1
+      return 0
+    }
+  )
+
+const iso2toDialCode = (iso2Value: string) => {
+  if(isEmpty(iso2Value)) {
+    return ''
+  }
+  return `+${findIso2(iso2Value)}`
+}
+
+export default class PhoneField extends PureComponent<PhoneFieldProps, PhoneFieldState> {
+
+  static defaultProps: Partial<PhoneFieldProps> = {
+    defaultPrefix: '',
+    defaultValue: '',
+    countriesWhiteList: [],
+    customCountriesName: {},
+    isInline: true,
+  }
+
+  state = {
+    countriesData: PhoneField.countriesData(
+      this.props.countriesWhiteList, this.props.customCountriesName
+    ),
+    countriesWhiteList: this.props.countriesWhiteList,
+    [FIELDS.PHONENUMBER]: this.props.defaultValue,
+    [FIELDS.PHONEPREFIX]: iso2toDialCode(this.props.defaultPrefix),
+    completePhoneNumber:
+    iso2toDialCode(this.props.defaultPrefix) + this.props.defaultValue,
+  }
+
+  onChange = ({ name, value }: OnChangeParameters) => {
+    this.setState({[name]: value}, () => {
+      this.props.onChange({
+        name: this.props.name,
+        value: {
+          [FIELDS.PHONENUMBER]: this.state[FIELDS.PHONENUMBER],
+          [FIELDS.PHONEPREFIX]: this.state[FIELDS.PHONEPREFIX],
+          completePhoneNumber:
+          (this.state[FIELDS.PHONEPREFIX] as string) +
+          (this.state[FIELDS.PHONENUMBER] as string),
```

Comment:
Not really happy to redefine `as string` here but don't find a better way.
It's cause of the allowance line 41 `[key: string]: any` in PhoneFieldState interface.

Reply:
that's weird, I'm going to pull this PR to check that